agent_api/src/main/java/dev/aikido/agent_api/storage/ServiceConfiguration.java

```diff
@@ -87,68 +85,39 @@ public boolean isIpBypassed(String ip) {
      * Check if the IP is blocked (e.g. Geo IP Restrictions)
      */
     public BlockedResult isIpBlocked(String ip) {
+        BlockedResult blockedResult = new BlockedResult(false, null);
+
         // Check for allowed ip addresses (i.e. only one country is allowed to visit the site)
         // Always allow access from private IP addresses (those include local IP addresses)
-        if (!allowedIps.isEmpty() && !isPrivateIp(ip)) {
-            boolean ipAllowed = false;
-            for (IPListEntry entry : allowedIps) {
-                if (entry.ipList.matches(ip)) {
-                    ipAllowed = true; // We allow IP addresses as long as they match with one of the lists.
-                    break;
-                }
-            }
-            if (!ipAllowed) {
-                return new BlockedResult(true, "not in allowlist");
-            }
+        if (!isPrivateIp(ip) && !firewallLists.matchesAllowedIps(ip)) {
+            blockedResult = new BlockedResult(true, "not in allowlist");
         }
 
         // Check for blocked ip addresses
-        for (IPListEntry entry : blockedIps) {
-            if (entry.ipList.matches(ip)) {
-                return new BlockedResult(true, entry.description);
+        for (ParsedFirewallLists.Match match : firewallLists.matchBlockedIps(ip)) {
+            StatisticsStore.incrementIpHits(match.key());
+            // when a blocking match is found, set blocked result if it hasn't been set already.
+            if (match.block() && !blockedResult.blocked()) {
+                blockedResult = new BlockedResult(true, match.description());
```

Comment:
I get that we're trying to avoid a second loop but returning when a block was found seems easier to reason about:

```java
public BlockedResult isIpBlocked(String ip) {
    // First check if IP is not in allowlist (unless it's private)
    if (!isPrivateIp(ip) && !firewallLists.matchesAllowedIps(ip)) {
        return new BlockedResult(true, "not in allowlist");
    }

    firewallLists.matchBlockedIps(ip).forEach(match -> 
        StatisticsStore.incrementIpHits(match.key())
    );

    for (ParsedFirewallLists.Match match : firewallLists.matchBlockedIps(ip)) {
        if (match.block()) {
            return new BlockedResult(true, match.description());
        }
    }
    
    return new BlockedResult(false, null);
}
```

Reply:
sure but this is "minimized" java code, the .forEach I tend not to use because we might want to port to older versions e.g. the "var" I just never use unless it's in tests, I think this is not really necessary as it's clear a response is being set, but I'll change it.